algebraic-graphs.cabal

```diff
@@ -67,10 +71,14 @@ library
                         Algebra.Graph.Relation.Reflexive,
                         Algebra.Graph.Relation.Symmetric,
                         Algebra.Graph.Relation.Transitive
-    build-depends:      array      >= 0.5 && < 0.8,
-                        base       >= 4.9 && < 5,
-                        containers >= 0.5 && < 0.8,
-                        deepseq    >= 1.4.2.0 && <1.5
+    build-depends:      array       >= 0.4     && < 0.6,
+                        base        >= 4.7     && < 5,
+                        base-compat >= 0.9.1   && < 0.10,
+                        containers  >= 0.5.5.1 && < 0.8,
+                        deepseq     >= 1.3.0.1 && < 1.5
+    if !impl(ghc >= 8.0)
+      build-depends:
+                        semigroups >=0.18.3 && <0.18.4
```

Comment:
I don't see where the `semigroups` package is used. Is it really needed?

Reply:
For `Data.Semigroup` with old `base`.

Reply:
Aha, I see. I will merge this as is for now, but I will try to get rid of the dependency on `semigroups` via CPP. Essentially, I would like to provide the `Semigroup` instance only if it is required by the new `Semigroup => Monoid` hierarchy. For older versions of `base` I think it's OK to provide only `Monoid`.

Reply:
I'd advice against that. Currently `semigroups` overhead is
- three lines in cabal file, and 
- few additional transitive dependencies for GHC < 8

However if you try to avoid that depedency at all cost:
- The API will be different across different GHC, which might cause *wtf* moments for users, as e.g. Hackage (and Stackage) haddocks will show `Semigroup` instance (as they are generated with recent GHC)
- there will be CPP clutter (guards for imports and instance itself at least).

---

Note: this is the case where we see 3-release-policy in practice. If you were to support only three latest GHC, with GHC-8.4 you could finally add `Semigroup` instance across whole support window`, without CPP *or* addtional dependencies. And in GHC-8.4 you have to, as `Semigroup` is a superclass of `Monoid` in `base-4.11` (GHC-8.4). FWIW, `algebraic-graphs` (this branch to be precise) compiles cleanly with quite recent GHC HEAD too, so we don't need to worry about this.

Reply:
I see, thanks! I agree that the API should be uniform across different GHC. The main obstacle in this case is that it is impossible to hide the `Semigroup` instance from the API.